File: algorithms/03_cpp/numerical/lagrange_interpolation.cpp
// Lagrange polynomial interpolation (Numerical Recipes 3.1).
#include <cassert>
#include <cmath>
#include <iostream>
#include <vector>

double lagrange(const std::vector<double> &xs, const std::vector<double> &ys, double x) {
    double total = 0.0;
    int n = static_cast<int>(xs.size());
    for (int i = 0; i < n; ++i) {
        double term = ys[i];
        for (int j = 0; j < n; ++j)
            if (i != j) term *= (x - xs[j]) / (xs[i] - xs[j]);
        total += term;
    }
    return total;
}

int main() {
    assert(std::abs(lagrange({0, 1, 2}, {1, 3, 2}, 1.5) - 2.875) < 1e-9);
    std::cout << "lagrange interpolation ok\n";
    return 0;
}

File: algorithms/07_typescript/math/lagrange_interpolation.ts
// Lagrange polynomial interpolation.
function lagrange(xs: number[], ys: number[], x: number): number {
  let total = 0;
  for (let i = 0; i < xs.length; i += 1) {
    let term = ys[i];
    for (let j = 0; j < xs.length; j += 1) {
      if (i !== j) term *= (x - xs[j]) / (xs[i] - xs[j]);
    }
    total += term;
  }
  return total;
}

const value = lagrange([0, 1, 2], [1, 3, 2], 1.5);
if (Math.abs(value - 2.875) > 1e-9) throw new Error("lagrange interpolation failed");
console.log(`P(1.5)=${value}`);
